backend/tests/TestUtils/TestUtils.fs

```diff
@@ -670,9 +670,15 @@ module Expect =
       eq ("matchCond" :: path) e e'
 
       NEList.iter2
-        (fun ((p, v) : MatchPattern * Expr) (p', v') ->
-          matchPatternEqualityBaseFn checkIDs path p p' errorFn
-          eq (string p :: path) v v')
+        (fun branch branch' ->
+          matchPatternEqualityBaseFn checkIDs path branch.pat branch'.pat errorFn
+          match branch.whenCondition, branch'.whenCondition with
+          | Some cond, Some cond' -> eq ("whenCondition" :: path) cond cond'
+          | None, None -> ()
+          | _ ->
+            errorFn ("whenCondition" :: path) (string actual) (string expected)
+            ()
+          eq (string branch.pat :: path) branch.rhs branch'.rhs)
```

Comment:
I am not sure what path is used for, could you please explain why are we using branch.pat instead of "rhs"?

Reply:
the `path` is used in error-reporting to show more exactly where 2 things weren't equal.
I think your intuiton is right, though, that the specific impl here is a bit weird.
I think what we really want is a path of `expr -> case [x] -> [part where it failed]`.

Here's an adjustment towards this that I think might help clarify

```
NEList.iteri2
  (fun i branch branch' ->
    let path = "Case {i}" :: path

    matchPatternEqualityBaseFn checkIDs ("pat" :: path) branch.pat branch'.pat errorFn

    match branch.whenCondition, branch'.whenCondition with
    | Some cond, Some cond' -> eq ("whenCondition" :: path) cond cond'
    | None, None -> ()
    | _ ->
      errorFn ("whenCondition" :: path) (string actual) (string expected)
      ()

    eq ("rhs" :: path) branch.rhs branch'.rhs)
  branches
  branches'
```

Reply:
maybe `let path = $"Case {i} - {branch.pat}" :: path` there would actually be a better compromise, so you have a choice of 'counting' the cases when debugging, or simply referring to whatever pattern is shown

Reply:
We should probably have a "make sure the 2 exprs have the same # of cases` check, right before the NEList.iter2 - see `check path (List.length typeArgs) (List.length typeArgs')` in `EApply` just above in that fn